asmcomp/cmmgen.ml

```diff
@@ -772,21 +777,39 @@ let bigarray_elt_size = function
   | Pbigarray_complex64 -> 16
 
 let bigarray_indexing unsafe elt_kind layout b args dbg =
-  let check_bound a1 a2 k =
-    if unsafe then k else Csequence(make_checkbound dbg [a1;a2], k) in
+  let check_bound bound idx v =
+    Csequence(make_checkbound dbg [bound;idx], v) in
+  (* Validates the given multidimensional offset against the array bounds and
+     transforms it into a one dimensional offset.  The offsets are expressions
+     evaluating to int. *)
   let rec ba_indexing dim_ofs delta_ofs = function
     [] -> assert false
   | [arg] ->
-      bind "idx" (untag_int arg)
-        (fun idx ->
-           check_bound (Cop(Cload Word,[field_address b dim_ofs])) idx idx)
+      (* Avoid the costly binding if [idx] is not needed *)
+      if unsafe then arg
+      else
+        bind "idx" arg (fun idx ->
+          (* Load the nativeint bound for the given dimension *)
+          let bound = Cop(Cload Word,[field_address b dim_ofs]) in
+          (* Transform the index to nativeint to compare it with the bound *)
+          let idxn = untag_int idx in
+          check_bound bound idxn idx)
   | arg1 :: argl ->
+      (* The remainder of the list is transformed into a one dimensional offset
+         *)
       let rem = ba_indexing (dim_ofs + delta_ofs) delta_ofs argl in
-      bind "idx" (untag_int arg1)
-        (fun idx ->
-          bind "bound" (Cop(Cload Word, [field_address b dim_ofs]))
-          (fun bound ->
-            check_bound bound idx (add_int (mul_int rem bound) idx))) in
+      (* Load the nativeint bound for the given dimension *)
+      let bound = Cop(Cload Word, [field_address b dim_ofs]) in
+      (* Avoid the costly bindings if [idx] and [bound] are not needed *)
+      if unsafe then add_int (mul_int (decr_int rem) bound) arg1
+      else
+        bind "idx" arg1 (fun idx ->
+          bind "bound" bound (fun bound ->
+            (* Transform the index to nativeint to compare it with the bound *)
+            let idxn = untag_int idx in
+            check_bound bound idxn
+              (add_int (mul_int (decr_int rem) bound) idx))) in
```

Comment:
Thanks for the extra comments. I find it very useful to have comments telling whether any given expression is expected to represent an OCaml tagged integer, and which represent a native, untagged integer -- this is one of the most perplexing aspect when reading code at this level. (I find "tagged" and "untagged" easier to follow than "int" and "nativeint", but maybe that's an existing convention in this part of the codebase.)

There are still things I do not understand in this code. Reading the patch with little knowledge of the context, here is what I see:

1. `check_bound` is unchanged
2. however, `check_bound` call-site in the easy case on line 796 is changed, with only the first index-like argument untagged (the second isn't anymore)
3. `let bound = ...` is unchanged
4. again, the call on lines 810-811 only untags the first argument.
5. the call also does something weird with the second one, turning `rem` into `decr_int` for no discernable purpose

My assumption, that explains (2) and (4), is that the specification of the `ba_indexing` function changed, it now returns a tagged integer instead of an untagged integer. While the comment before `ba_indexing` in fact says that the function returns a tagged integer (but it took me a few re-reading to understand that it said that), the fact that this is a change is not at all clear. Of course, comments should in general not discuss the previous state of the code; this is what the commit message is for.

That still doesn't explain much of (5). I guess that you use the fact that tagging/untagging through a multiplication cancels out to `decr_int`, but this part specifically is too tricky to review for someone not familiar (or interested) in bit-level tricks without a comment pointing out which equation it is.

Finally, here are two small changes that I think would make the whole thing slightly easier to follow:
- a comment explaining the input/output tagedness of `array_indexing` and `bigarray_indexing`
- renaming the `check_bound` local function to `check_ba_bound`, to make grepping for its use and definition easier to not confuse with the other (semantically different) `check_bound` function in this module.